Src/VimCore/CommonOperations.fs

```diff
@@ -345,9 +345,15 @@ type internal CommonOperations
                     None
                 elif caretLineNumber < (topLineNumber + _globalSettings.ScrollOffset) || caretLineNumber > (bottomLineNumber - _globalSettings.ScrollOffset) then
                     let lineNumber = caretLineNumber
-                    let lineNumber = max lineNumber (topLineNumber + _globalSettings.ScrollOffset)
-                    let lineNumber = min lineNumber (bottomLineNumber - _globalSettings.ScrollOffset)
-                    Some lineNumber
+                    let topOffset = min caretLineNumber _globalSettings.ScrollOffset
+                    let lastVisualLineNumber = _textView.VisualSnapshot.LineCount - 1
+                    let bottomOffset = min (lastVisualLineNumber - caretLineNumber) _globalSettings.ScrollOffset
+                    let lineNumber = max lineNumber (topLineNumber + topOffset)
+                    let lineNumber = min lineNumber (bottomLineNumber - bottomOffset)
+                    if lineNumber <> caretLineNumber then
+                        Some lineNumber
+                    else
+                        None

```

Comment:
Why did you choose to implement this change here vs. directly in `CommandUtil.ScrollLines`? The latter is where I would have thought to do this. Doing it here seems like it would affect `CommandUtil.ScrollPages` as well.

Reply:
`ScrollLines` was already putting the caret on the right line and then it calls this function which is named `AdjustCaretForScrollOffset`. So anybody should be able to call `AdjustCaretForScrollOffset` and expect that it will only adjust the caret when the caret is "too close to the top/bottom of the window, except when at the top/bottom of the buffer."

In fact,`ScrollPages` didn't have the bug because of the way that scrolling up leaves the caret at the bottom of the screen and vice versa. However, if it didn't it would also need to correctly "not move the caret away from the edge" when the edge is in fact the top/bottom of the buffer.